pages/api/zapbye.ts

```diff
@@ -0,0 +1,100 @@
+            salesforceConnection = new jsforce.Connection();
+            await salesforceConnection.login(CONSTANTS.SALESFORCE_EMAIL, CONSTANTS.SALESFORCE_PASSWORD + CONSTANTS.SALESFORCE_SECURITY_TOKEN, function(err, res) {
+                  if (err) {
+                        salesforceConnection = null;
+                        res.statusCode = 500;
+                        res.body = err;
+                        return console.error(err);
+                  }
+            });
+      }
+      let lemlistPayload = req.body;
+      await salesforceConnection.sobject(CONSTANTS.SALESFORCE_SOBJECT_TYPE).create({
+            TaskSubtype: CONSTANTS.SALESFORCE_TASK_SUBTYPE,
+            Subject: CONSTANTS.SALESFORCE_TASK_SUBJECT,
+            Status: CONSTANTS.SALESFORCE_STATUS,
+            RecordTypeId: CONSTANTS.SALESFORCE_LEMLISTTASK_ID,
+            ActivityDate: new Date(), //DueDate
+            Lemlist_Campaign_Name__c: await getLemlistCampaignNameFromCampaignId(lemlistPayload.campaignId),
+            Lemlist_Sender__c: lemlistPayload.leadFirstName + ' ' + lemlistPayload.leadLastName,
+            Lemlist_Email__c: lemlistPayload.leadEmail,
+            Lemlist_Task_Type__c: lemlistPayload.type,
+            Lemlist_Sequence_Step_Number__c: lemlistPayload.sequenceStep,
+      },{},function(err, ret) {
+            if (err || !ret.success) {
+                  return console.error(err, ret);
+            }
+            console.log('Task created successfully: ', ret);
+      });
+      res.statusCode = 200;
+      res.json({ response: "200 OK" });
+}
+
+interface LemlistCampaign {
+      _id: string,
+      name: string,
+      archived: boolean
+}
+
+async function getLemlistCampaigns(): Promise<JSON> {
+      let endpoint = "https://:" + CONSTANTS.LEMLIST_API_KEY + "@api.lemlist.com/api/campaigns";
+      console.log("Attempting to fetch Lemlist Campaigns from API");
+      const response: Response = await fetch(endpoint);
+      return response.json();
+}
+
+async function getLemlistCampaignNameFromCampaignId(campaignId: string): Promise<string> {
+      let campaign = filterCampaignById(cachedCampaignsArray, campaignId);
+      if(!campaign) {
+            //we keep track of how long ago we last pulled from Lemlist so we don't end up getting rate limited if there's an ID we can't find
+            let now = Date.now();
+            if(!lastCampaignCheckedTimestamp || now - lastCampaignCheckedTimestamp > CONSTANTS.LEMLIST_CAMPAIGN_FETCH_COOLDOWN_TIME_IN_MS) {
+                  cachedCampaignsArray = await getLemlistCampaigns() as unknown as Array<LemlistCampaign>;
+                  campaign = filterCampaignById(cachedCampaignsArray, campaignId);
+                  lastCampaignCheckedTimestamp = now;
+            }
+      }
+      return campaign;
+}
+
+function filterCampaignById(campaigns: Array<LemlistCampaign>, campaignId: string): string {
+      for (let i = 0; i < campaigns.length; i++) {
```

Comment:
(Guard suggestion) I think `campaigns` will be 'undefined' in the first run since `cachedCampaignsArray` is referenced on line  80, but actually populated on line 85. 

```suggestion
      for (let i = 0; i < (campaigns || []).length; i++) {
```

Reply:
Good catch. I'll pre-init to an empty array when declaring.

Reply:
Additionally you can use a (for .. in) loop here :)

Reply:
@mattvv Yeah this was me not wanting to bother figuring out what foreach syntax was, but I'm gonna switch it over since I don't really care about the index itself.